helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -703,8 +939,6 @@ public void process(WatchedEvent event) {
         if (event.getState() == KeeperState.Expired) {
           getEventLock().getZNodeEventCondition().signalAll();
           getEventLock().getDataChangedCondition().signalAll();
-          // We also have to notify all listeners that something might have changed
-          fireAllEvents();
```

Comment:
It is in the finally block. With this change, when getShutdownTrigger() == true, some notifications might be missed.